File: jdroid-java-core/src/main/java/com/jdroid/java/gson/MethodSerializer.java
package com.jdroid.java.gson;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.lang.reflect.Type;

public class MethodSerializer implements JsonSerializer<Object> {

	private GsonBuilder gsonBuilder;

	public MethodSerializer(GsonBuilder gsonBuilder) {
		this.gsonBuilder = gsonBuilder;
	}

	public MethodSerializer() {
		gsonBuilder = GsonBuilderFactory.createGsonBuilder();
	}

	public JsonElement serialize(Object src, Type typeOfSrc, JsonSerializationContext context) {
		Gson gson = gsonBuilder.create();

		JsonObject tree = (JsonObject)gson.toJsonTree(src);

		try {
			PropertyDescriptor[] properties = Introspector.getBeanInfo(src.getClass()).getPropertyDescriptors();
			for (PropertyDescriptor property : properties) {
				if (property.getReadMethod().getAnnotation(ExposeMethod.class) != null) {
					Object result = property.getReadMethod().invoke(src, (Object[])null);
					tree.add(property.getName(), gson.toJsonTree(result));
				}
			}
		} catch (Exception e) {
			throw new RuntimeException(e);
		}

		return tree;
	}
}

File: jdroid-java-core/src/main/java/com/dipien/java/gson/MethodSerializer.kt
package com.dipien.java.gson

import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.beans.Introspector
import java.lang.reflect.Type

class MethodSerializer : JsonSerializer<Any> {

    private var gsonBuilder: GsonBuilder? = null

    constructor(gsonBuilder: GsonBuilder) {
        this.gsonBuilder = gsonBuilder
    }

    constructor() {
        gsonBuilder = GsonBuilderFactory.createGsonBuilder()
    }

    override fun serialize(src: Any, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val gson = gsonBuilder!!.create()

        val tree = gson.toJsonTree(src) as JsonObject

        try {
            val properties = Introspector.getBeanInfo(src.javaClass).propertyDescriptors
            for (property in properties) {
                if (property.readMethod.getAnnotation(ExposeMethod::class.java) != null) {
                    val result = property.readMethod.invoke(src, null as Array<Any>?)
                    tree.add(property.name, gson.toJsonTree(result))
                }
            }
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        return tree
    }
}
